File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import com.ctre.phoenix6.StatusCode;
import com.ctre.phoenix6.configs.CurrentLimitsConfigs;
import com.ctre.phoenix6.configs.MotionMagicConfigs;
import com.ctre.phoenix6.configs.MotorOutputConfigs;
import com.ctre.phoenix6.configs.Slot0Configs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.signals.NeutralModeValue;

import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

import frc.robot.Constants;
import frc.robot.frc8011.WayiMotor;
import frc.robot.logging.MotorHealthConfig;
import frc.robot.logging.RobotHealthLogger;
import frc.robot.shooting.ShotSetpoint;

public class ShooterSubsystem extends SubsystemBase {
    public enum PitchHomeState {
        UNHOMED,
        HOMING,
        HOMED,
        FAULT
    }

    private static final double IDLE_SPEED_RPS = 0.0;
    private static final double PITCH_HOME_TIMEOUT_SECONDS = 2.0;
    private static final double PITCH_HOME_CURRENT_AMPS = -10.0;
    private static final double PITCH_HOME_CURRENT_THRESHOLD_AMPS = -8.0;
    private static final int PITCH_HOME_CURRENT_COUNT = 3;
    private static final double PITCH_GEAR_RATIO = 18.67;

    private final WayiMotor shooterLeftDown = new WayiMotor(
            Constants.ShooterConstants.shooterLeftDownMotorId,
            Constants.CanConstants.rioCanBus);
    private final WayiMotor shooterLeftUp = new WayiMotor(
            Constants.ShooterConstants.shooterLeftUpMotorId,
            Constants.CanConstants.rioCanBus);
    private final WayiMotor shooterRightUp = new WayiMotor(
            Constants.ShooterConstants.shooterRightUpMotorId,
            Constants.CanConstants.rioCanBus);
    private final WayiMotor shooterRightDown = new WayiMotor(
            Constants.ShooterConstants.shooterRightDownMotorId,
            Constants.CanConstants.rioCanBus);
    private final WayiMotor shooterPitch = new WayiMotor(
            Constants.ShooterConstants.shooterPitchMotorId,
            Constants.CanConstants.rioCanBus);

    private ShotSetpoint targetSetpoint = new ShotSetpoint(15.0, 0.0, 0.2);
    private boolean shotActive = false;
    private boolean flywheelFollowersArmed = false;
    private PitchHomeState pitchHomeState = PitchHomeState.UNHOMED;
    private int pitchHomeCurrentCounter = 0;
    private final Timer pitchHomeTimer = new Timer();
    private double healthFlywheelVelocity;
    private double healthPitchAngle;

    public ShooterSubsystem() {
        initialize();
    }

    @Override
    public void periodic() {
        shooterRightUp.receiveVelocity();
        healthFlywheelVelocity = shooterRightUp.getCachedVelocity();

        if (!DriverStation.isEnabled()) {
            flywheelFollowersArmed = false;
            shooterRightUp.setCoast();
            shooterRightUp.control();
            shooterPitch.setBrake();
            shooterPitch.control();
            pitchHomeTimer.stop();
            pitchHomeCurrentCounter = 0;
            pitchHomeState = PitchHomeState.UNHOMED;
            return;
        }

        if (pitchHomeState == PitchHomeState.UNHOMED) {
            beginPitchHoming();
        }

        if (pitchHomeState == PitchHomeState.HOMING) {
            runPitchHoming();
        } else if (pitchHomeState == PitchHomeState.HOMED) {
            setShootPitchAngle(targetSetpoint.pitchDeg());
            shooterPitch.control();
        } else {
            shooterPitch.setBrake();
            shooterPitch.control();
        }

        shooterRightUp.setVelocityTorqueCurrent(
                shotActive ? targetSetpoint.flywheelRps() : IDLE_SPEED_RPS);
        shooterRightUp.control();
        if (!flywheelFollowersArmed) {
            commandFlywheelFollowers();
            flywheelFollowersArmed = true;
        }
        healthPitchAngle = getPitchAngle();

        SmartDashboard.putNumber(
                "Shooting/FlywheelTargetRps",
                shotActive ? targetSetpoint.flywheelRps() : IDLE_SPEED_RPS);
        SmartDashboard.putNumber("Shooting/FlywheelActualRps", healthFlywheelVelocity);
        SmartDashboard.putNumber("Shooting/PitchTargetDeg", targetSetpoint.pitchDeg());
        SmartDashboard.putNumber("Shooting/PitchActualDeg", healthPitchAngle);
        SmartDashboard.putNumber("Shooting/PitchHomeState", pitchHomeState.ordinal());
    }

    private void initialize() {
        TalonFXConfiguration flywheelConfig = new TalonFXConfiguration()
                .withMotorOutput(new MotorOutputConfigs()
                        .withInverted(InvertedValue.CounterClockwise_Positive)
                        .withNeutralMode(NeutralModeValue.Coast))
                .withCurrentLimits(new CurrentLimitsConfigs()
                        .withStatorCurrentLimit(120)
                        .withStatorCurrentLimitEnable(true)
                        .withSupplyCurrentLimit(50)
                        .withSupplyCurrentLimitEnable(true))
                .withSlot0(new Slot0Configs()
                        .withKS(4.875)
                        .withKP(9));
        applyWithRetry(shooterLeftDown, flywheelConfig);
        applyWithRetry(shooterLeftUp, flywheelConfig);
        applyWithRetry(shooterRightUp, flywheelConfig);
        applyWithRetry(shooterRightDown, flywheelConfig);
        shooterRightUp.setInvert(-1);

        shooterLeftDown.setFollower(shooterRightUp.getData().deviceId, true);
        shooterLeftUp.setFollower(shooterRightUp.getData().deviceId, true);
        shooterRightDown.setFollower(shooterRightUp.getData().deviceId, false);
        commandFlywheelFollowers();

        TalonFXConfiguration pitchConfig = new TalonFXConfiguration()
                .withMotorOutput(new MotorOutputConfigs()
                        .withInverted(InvertedValue.CounterClockwise_Positive)
                        .withNeutralMode(NeutralModeValue.Brake))
                .withCurrentLimits(new CurrentLimitsConfigs()
                        .withStatorCurrentLimit(60)
                        .withStatorCurrentLimitEnable(true)
                        .withSupplyCurrentLimit(20)
                        .withSupplyCurrentLimitEnable(true))
                .withSlot0(new Slot0Configs()
                        .withKP(7)
                        .withKI(0.8)
                        .withKD(0.02))
                .withMotionMagic(new MotionMagicConfigs()
                        .withMotionMagicCruiseVelocity(0)
                        .withMotionMagicExpo_kV(0.1)
                        .withMotionMagicExpo_kA(0.01));
        applyWithRetry(shooterPitch, pitchConfig);
        shooterPitch.setGearRatio(PITCH_GEAR_RATIO);
        shooterPitch.setInvert(-1);
        shooterPitch.setPhysicalLimits(
                0.0,
                Constants.ShooterConstants.pitchMotorMaxPositionRot / PITCH_GEAR_RATIO,
                120.0 / PITCH_GEAR_RATIO,
                40.0);
        shooterPitch.setCurrentSpeed(0.1);
        shooterPitch.setStatusSignalUpdateFrequency(50);
    }

    public void applyShotSetpoint(ShotSetpoint setpoint) {
        targetSetpoint = setpoint;
        shotActive = true;
    }

    public void setIdle() {
        shotActive = false;
        targetSetpoint = new ShotSetpoint(IDLE_SPEED_RPS, 0.0, 0.2);
    }

    public void stop() {
        shotActive = false;
        shooterRightUp.setCoast();
        shooterRightUp.control();
    }

    public Command stopCommand() {
        return runOnce(this::stop);
    }

    public void beginPitchHoming() {
        if (!DriverStation.isEnabled()) {
            pitchHomeState = PitchHomeState.UNHOMED;
            return;
        }
        if (RobotBase.isSimulation()) {
            targetSetpoint = new ShotSetpoint(targetSetpoint.flywheelRps(), targetSetpoint.feederRps(), 0.0);
            pitchHomeState = PitchHomeState.HOMED;
            return;
        }
        pitchHomeState = PitchHomeState.HOMING;
        pitchHomeCurrentCounter = 0;
        pitchHomeTimer.restart();
    }

    public PitchHomeState getPitchHomeState() {
        return pitchHomeState;
    }

    public boolean isPitchHomed() {
        return pitchHomeState == PitchHomeState.HOMED;
    }

    public double getShootVelocity() {
        return shooterRightUp.getVelocity();
    }

    public double getPitchAngle() {
        double normalized = shooterPitch.getNormalizedPosition();
        return normalized
                * (Constants.ShooterConstants.maxPitchAngleDeg
                        - Constants.ShooterConstants.minPitchAngleDeg)
                + Constants.ShooterConstants.minPitchAngleDeg;
    }

    public boolean isFlywheelReady(double toleranceRps) {
        return shotActive
                && Math.abs(shooterRightUp.getVelocity() - targetSetpoint.flywheelRps()) <= toleranceRps;
    }

    public boolean isPitchReady(double toleranceDeg) {
        return isPitchHomed()
                && Math.abs(getPitchAngle() - targetSetpoint.pitchDeg()) <= toleranceDeg;
    }

    public ShotSetpoint getTargetSetpoint() {
        return targetSetpoint;
    }

    /** Registers the shooter hardware and control context with the central logger. */
    public void registerHealthLogging(RobotHealthLogger logger) {
        if (logger == null) {
            return;
        }
        int flywheelLeaderId = shooterRightUp.getData().deviceId;
        logger.registerTalonFX(
                "Shooter",
                "LeftDown",
                shooterLeftDown.getMotor(),
                MotorHealthConfig.follower(flywheelLeaderId, true));
        logger.registerTalonFX(
                "Shooter",
                "LeftUp",
                shooterLeftUp.getMotor(),
                MotorHealthConfig.follower(flywheelLeaderId, true));
        logger.registerTalonFX(
                "Shooter",
                "RightUpLeader",
                shooterRightUp.getMotor(),
                MotorHealthConfig.leader());
        logger.registerTalonFX(
                "Shooter",
                "RightDown",
                shooterRightDown.getMotor(),
                MotorHealthConfig.follower(flywheelLeaderId, false));
        logger.registerTalonFX("Shooter", "Pitch", shooterPitch.getMotor());
        logger.registerSubsystem(
                "Shooter",
                this,
                this::isHealthActive,
                this::getHealthState,
                () -> shotActive ? targetSetpoint.flywheelRps() : IDLE_SPEED_RPS,
                () -> healthFlywheelVelocity);
    }

    private boolean isHealthActive() {
        return shotActive || pitchHomeState == PitchHomeState.HOMING;
    }

    private String getHealthState() {
        if (!DriverStation.isEnabled()) {
            return "Disabled";
        }
        return switch (pitchHomeState) {
            case UNHOMED -> "Unhomed";
            case HOMING -> "Homing";
            case FAULT -> "Fault";
            case HOMED -> {
                if (!shotActive) {
                    yield "Idle";
                }
                yield isHealthFlywheelReady(1.0) && isHealthPitchReady(1.0)
                        ? "Ready"
                        : "SpinningUp";
            }
        };
    }

    private boolean isHealthFlywheelReady(double toleranceRps) {
        return shotActive
                && Math.abs(healthFlywheelVelocity - targetSetpoint.flywheelRps())
                        <= toleranceRps;
    }

    private boolean isHealthPitchReady(double toleranceDeg) {
        return isPitchHomed()
                && Math.abs(healthPitchAngle - targetSetpoint.pitchDeg()) <= toleranceDeg;
    }

    private void runPitchHoming() {
        shooterPitch.setCurrent(PITCH_HOME_CURRENT_AMPS);
        shooterPitch.control();

        if (shooterPitch.getCurrent() < PITCH_HOME_CURRENT_THRESHOLD_AMPS) {
            pitchHomeCurrentCounter++;
        } else {
            pitchHomeCurrentCounter = 0;
        }

        if (pitchHomeCurrentCounter >= PITCH_HOME_CURRENT_COUNT) {
            shooterPitch.reset(shooterPitch.getAbsPosition());
            shooterPitch.setBrake();
            shooterPitch.control();
            targetSetpoint = new ShotSetpoint(targetSetpoint.flywheelRps(), targetSetpoint.feederRps(), 0.0);
            pitchHomeState = PitchHomeState.HOMED;
            pitchHomeTimer.stop();
            return;
        }

        if (pitchHomeTimer.hasElapsed(PITCH_HOME_TIMEOUT_SECONDS)) {
            shooterPitch.setBrake();
            shooterPitch.control();
            pitchHomeState = PitchHomeState.FAULT;
            pitchHomeTimer.stop();
        }
    }

    private void setShootPitchAngle(double targetAngleDeg) {
        double clamped = Math.max(
                Constants.ShooterConstants.minPitchAngleDeg,
                Math.min(Constants.ShooterConstants.maxPitchAngleDeg, targetAngleDeg));
        double normalized = (clamped - Constants.ShooterConstants.minPitchAngleDeg)
                / (Constants.ShooterConstants.maxPitchAngleDeg
                        - Constants.ShooterConstants.minPitchAngleDeg);
        shooterPitch.setNormalizedMotionPosition(normalized);
    }

    private void commandFlywheelFollowers() {
        shooterLeftDown.control();
        shooterLeftUp.control();
        shooterRightDown.control();
    }

    private static void applyWithRetry(WayiMotor motor, TalonFXConfiguration config) {
        for (int i = 0; i < 5; i++) {
            StatusCode status = motor.applyConfig(config);
            if (status.isOK()) {
                return;
            }
        }
    }
}

File: cpp/src/main/cpp/subsystems/ShooterSubsystem.cpp
#include "subsystems/ShooterSubsystem.h"

#include <algorithm>
#include <cmath>

#include <frc/DriverStation.h>
#include <frc/RobotBase.h>
#include <frc/smartdashboard/SmartDashboard.h>

using namespace subsystems;
using namespace units::literals;

void ShooterSubsystem::Initialization() {
  configs::TalonFXConfiguration flywheel_config{};
  flywheel_config.MotorOutput.Inverted = 0;
  flywheel_config.MotorOutput.NeutralMode = 0;
  flywheel_config.CurrentLimits.StatorCurrentLimit = 120_A;
  flywheel_config.CurrentLimits.StatorCurrentLimitEnable = true;
  flywheel_config.CurrentLimits.SupplyCurrentLimit = 50_A;
  flywheel_config.CurrentLimits.SupplyCurrentLimitEnable = true;

  configs::Slot0Configs& flywheel_slot0 = flywheel_config.Slot0;
  flywheel_slot0.kS = 4.875;
  flywheel_slot0.kP = 9;

  shooter_right_up_.Applyconfig(flywheel_config);
  shooter_right_up_.setinvert(-1);

  shooter_left_down_.setfollowControl(shooter_right_up_.Getdata().deviceId,
                                       true);
  shooter_left_up_.setfollowControl(shooter_right_up_.Getdata().deviceId,
                                     true);
  shooter_right_down_.setfollowControl(shooter_right_up_.Getdata().deviceId,
                                        false);
  shooter_left_down_.Control();
  shooter_left_up_.Control();
  shooter_right_down_.Control();

  configs::TalonFXConfiguration pitch_config{};
  pitch_config.MotorOutput.Inverted = 0;
  pitch_config.MotorOutput.NeutralMode = 1;
  pitch_config.CurrentLimits.StatorCurrentLimit = 60_A;
  pitch_config.CurrentLimits.StatorCurrentLimitEnable = true;
  pitch_config.CurrentLimits.SupplyCurrentLimit = 20_A;
  pitch_config.CurrentLimits.SupplyCurrentLimitEnable = true;

  configs::Slot0Configs& pitch_slot0 = pitch_config.Slot0;
  pitch_slot0.kP = 7;
  pitch_slot0.kI = 0.8;
  pitch_slot0.kD = 0.02;

  configs::MotionMagicConfigs& pitch_motion_magic = pitch_config.MotionMagic;
  pitch_motion_magic.MotionMagicCruiseVelocity = 0_tps;
  pitch_motion_magic.MotionMagicExpo_kV = 0.1_V / 1_tps;
  pitch_motion_magic.MotionMagicExpo_kA = 0.01_V / 1_tr_per_s_sq;

  shooter_pitch_.Applyconfig(pitch_config);
  shooter_pitch_.setgearRatio(18.67);
  shooter_pitch_.setinvert(-1);
  shooter_pitch_.setPhysicalLimits(
      0, ShooterConstants::kPitchMotorMaxposition / 18.67, 120 / 18.67, 40);
  shooter_pitch_.setCurrent_Speed(0.1);
  shooter_pitch_.SetStatusSignalUpdateFrequency(50_Hz);
}

void ShooterSubsystem::Periodic() {
  shooter_right_up_.ReceiveVelocity();

  if (!frc::DriverStation::IsEnabled()) {
    shooter_right_up_.setcoast();
    shooter_right_up_.Control();
    shooter_pitch_.setbrake();
    shooter_pitch_.Control();
    pitch_home_timer_.Stop();
    pitch_home_current_counter_ = 0;
    pitch_home_state_ = PitchHomeState::kUnhomed;
    return;
  }

  if (pitch_home_state_ == PitchHomeState::kUnhomed) {
    BeginPitchHoming();
  }

  if (pitch_home_state_ == PitchHomeState::kHoming) {
    RunPitchHoming();
  } else if (pitch_home_state_ == PitchHomeState::kHomed) {
    SetShootPitchAngle(target_setpoint_.pitch);
    shooter_pitch_.Control();
  } else {
    shooter_pitch_.setbrake();
    shooter_pitch_.Control();
  }

  shooter_right_up_.setvelocitytorquecurrent(
      shot_active_ ? target_setpoint_.flywheel.value() : kIdleSpeed.value());
  shooter_right_up_.Control();

  frc::SmartDashboard::PutNumber("Shooting/FlywheelTargetRps",
                                 shot_active_ ? target_setpoint_.flywheel.value()
                                              : kIdleSpeed.value());
  frc::SmartDashboard::PutNumber("Shooting/FlywheelActualRps",
                                 GetShootVelocity());
  frc::SmartDashboard::PutNumber("Shooting/PitchTargetDeg",
                                 target_setpoint_.pitch.value());
  frc::SmartDashboard::PutNumber("Shooting/PitchActualDeg", GetPitchAngle());
  frc::SmartDashboard::PutNumber(
      "Shooting/PitchHomeState", static_cast<int>(pitch_home_state_));
}

void ShooterSubsystem::ApplyShotSetpoint(
    const shooting::ShotSetpoint& setpoint) {
  target_setpoint_ = setpoint;
  shot_active_ = true;
}

void ShooterSubsystem::SetIdle() {
  shot_active_ = false;
  target_setpoint_.flywheel = kIdleSpeed;
  target_setpoint_.pitch = 0.2_deg;
}

void ShooterSubsystem::Stop() {
  shot_active_ = false;
  shooter_right_up_.setcoast();
  shooter_right_up_.Control();
}

frc2::CommandPtr ShooterSubsystem::StopCommandPtr() {
  return this->RunOnce([this] { Stop(); });
}

void ShooterSubsystem::BeginPitchHoming() {
  if (!frc::DriverStation::IsEnabled()) {
    pitch_home_state_ = PitchHomeState::kUnhomed;
    return;
  }
  if (frc::RobotBase::IsSimulation()) {
    target_setpoint_.pitch = 0_deg;
    pitch_home_state_ = PitchHomeState::kHomed;
    return;
  }
  pitch_home_state_ = PitchHomeState::kHoming;
  pitch_home_current_counter_ = 0;
  pitch_home_timer_.Restart();
}

void ShooterSubsystem::RunPitchHoming() {
  shooter_pitch_.setcurrent(-10.0);
  shooter_pitch_.Control();

  if (shooter_pitch_.GetCurrent() < -8.0) {
    ++pitch_home_current_counter_;
  } else {
    pitch_home_current_counter_ = 0;
  }

  if (pitch_home_current_counter_ >= 3) {
    shooter_pitch_.Reset(shooter_pitch_.GetAbsPosition());
    shooter_pitch_.setbrake();
    shooter_pitch_.Control();
    target_setpoint_.pitch = 0_deg;
    pitch_home_state_ = PitchHomeState::kHomed;
    pitch_home_timer_.Stop();
    return;
  }

  if (pitch_home_timer_.HasElapsed(kPitchHomeTimeout)) {
    shooter_pitch_.setbrake();
    shooter_pitch_.Control();
    pitch_home_state_ = PitchHomeState::kFault;
    pitch_home_timer_.Stop();
  }
}

double ShooterSubsystem::GetShootVelocity() {
  return shooter_right_up_.GetVelocity();
}

double ShooterSubsystem::GetPitchAngle() {
  const double normalized = shooter_pitch_.GetNormalizedPosition();
  return normalized *
         (ShooterConstants::kMaxPitchAngle - ShooterConstants::kMinPitchAngle) +
         ShooterConstants::kMinPitchAngle;
}

bool ShooterSubsystem::IsFlywheelReady(
    units::turns_per_second_t tolerance) {
  const double actual = shooter_right_up_.GetVelocity();
  return shot_active_ &&
         std::abs(actual - target_setpoint_.flywheel.value()) <=
             tolerance.value();
}

bool ShooterSubsystem::IsPitchReady(units::degree_t tolerance) {
  return IsPitchHomed() &&
         std::abs(GetPitchAngle() - target_setpoint_.pitch.value()) <=
             tolerance.value();
}

void ShooterSubsystem::SetShootPitchAngle(units::degree_t target_angle) {
  const double clamped =
      std::clamp(target_angle.value(), ShooterConstants::kMinPitchAngle,
                 ShooterConstants::kMaxPitchAngle);
  const double normalized =
      (clamped - ShooterConstants::kMinPitchAngle) /
      (ShooterConstants::kMaxPitchAngle - ShooterConstants::kMinPitchAngle);
  shooter_pitch_.setNormalizedMotionPosition(normalized);
}
